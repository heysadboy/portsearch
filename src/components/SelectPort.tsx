// This component can be used to select the option from a given list.
// You can see the suggestions as you type.

import { useState, FC } from "react";
import { IPort } from "../types";

interface ISelectPort {
    ports: IPort[],
    placeholder: string,
    onSelect: (port: string) => void
}

const SelectPort: FC<ISelectPort> = ({ ports, placeholder, onSelect }) => {
    //Controlled component to set the input
    const [input, setInput] = useState("");

    const setSearchInput = (e: any) => {
        setInput(e.target.innerText);
        onSelect(e.target.dataset.value);
    }

    //Render search options as the user types in the list
    const searchListOptions = ports
        .filter((port) => {
            //Returning suggestion based on the match with name and code
            return port.name.toLowerCase().startsWith(input.toLowerCase()) || port.code.toLowerCase().startsWith(input.toLowerCase())
        })
        .map((ports) => {
            return <div className="item search-option" key={ports.code} data-value={ports.code} onClick={setSearchInput} >{`${ports.name} (${ports.code})`}</div>
        });

    const onInputChange = (e: any) => {
        setInput(e.target.value);
    }

    return (
        <div>
            <div className="ui icon transparent input select-input">
                <input type="text" value={input} onChange={onInputChange} placeholder={placeholder} />
                <i className="globe icon"></i>
            </div>
            {
                input &&
                <div className="ui relaxed selection list search-list">
                    {searchListOptions}
                </div>
            }
        </div>
    );
}

export default SelectPort;