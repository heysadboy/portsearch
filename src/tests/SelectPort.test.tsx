import TestRenderer from "react-test-renderer";
import SelectPort from "../components/SelectPort";

const ports = [
    {
        code: "CNSGH",
        name: "Shanghai"
    },
    {
        code: "NOOSL",
        name: "Oslo"
    }
]

test("SelectPort Component Testing.", () => {
    const component = TestRenderer.create(
        <SelectPort ports={ports} placeholder="test_placeholder" onSelect={() => {}} />
    );

    let tree: any = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(component.root.findByType("input").props.placeholder).toEqual("test_placeholder");
});