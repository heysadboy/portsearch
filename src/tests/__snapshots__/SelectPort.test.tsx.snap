// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectPort Component Testing. 1`] = `
<div>
  <div
    className="ui icon transparent input select-input"
  >
    <input
      onChange={[Function]}
      placeholder="test_placeholder"
      type="text"
      value=""
    />
    <i
      className="globe icon"
    />
  </div>
  
</div>
`;
